packages/hebao_v1/test/testGuardians.ts

```diff
@@ -452,5 +452,178 @@ contract("GuardiansModule", (accounts: string[]) => {
         );
       }
     );
+
+    it(
+      description(
+        "owner should not be able to confirm guardian removal after expired"
+      ),
+      async () => {
+        useMetaTx = metaTx;
+        const owner = ctx.owners[0];
+        const { wallet } = await createWallet(ctx, owner);
+        const group = 0;
+
+        // The first guardian is added immediately (so cannot be cancelled)
+        await addGuardianChecked(owner, wallet, ctx.guardians[0], group);
+
+        // Now remove the guardian
+        await executeTransaction(
+          ctx.guardianModule.contract.methods.removeGuardian(
+            wallet,
+            ctx.guardians[0]
+          ),
+          ctx,
+          useMetaTx,
+          wallet,
+          [owner],
+          { from: owner }
+        );
+
+        // Skip forward `pendingPeriod` + `confirmPeriod` seconds
+        await advanceTimeAndBlockAsync(pendingPeriod + confirmPeriod);
+
+        // Try to confirm
+        await expectThrow(
+          executeTransaction(
+            ctx.guardianModule.contract.methods.confirmGuardianRemoval(
+              wallet,
+              ctx.guardians[0]
+            ),
+            ctx,
+            useMetaTx,
+            wallet,
+            [owner],
+            { from: owner }
+          ),
+          "TOO_EARLY_OR_EXPIRED"
+        );
+
+        // Now cancel
+        await executeTransaction(
+          ctx.guardianModule.contract.methods.cancelGuardianRemoval(
+            wallet,
+            ctx.guardians[0]
+          ),
+          ctx,
+          useMetaTx,
+          wallet,
+          [owner],
+          { from: owner }
+        );
+
+        // Make sure the guardian is not removed
+        assert(
+          await ctx.securityStore.isGuardian(wallet, ctx.guardians[0]),
+          "should be guardian"
+        );
+      }
+    );
+
+
   });
+
+  it(
+    description(
+      "anyone should be able to confirm guardian addition"
+    ),
+    async () => {
+      useMetaTx = false;
+      const owner = ctx.owners[0];
+      const { wallet } = await createWallet(ctx, owner);
+      const group = 0;
+
+      // The first guardian is added immediately (so cannot be cancelled)
+      await addGuardianChecked(owner, wallet, ctx.guardians[0], group);
+
+      // Add the second guardian which is added after a delay
+      await executeTransaction(
+        ctx.guardianModule.contract.methods.addGuardian(
+          wallet,
+          ctx.guardians[1],
+          group
+        ),
+        ctx,
+        useMetaTx,
+        wallet,
+        [owner],
+        { from: owner }
+      );
+
+      // Skip forward `pendingPeriod` + 1 seconds
+      await advanceTimeAndBlockAsync(pendingPeriod + 1);
+
+      const user = accounts[10];
+
+      // Try to confirm
+      await executeTransaction(
```

Comment:
This will use a meta tx signed by the wallet owner to confirm the guardian, so this isn't really a test that anyone can confirm the guardian addition. I think a better way to test is is to call `confirmGuardianAddition` immediately on the module instead of via a meta tx.

A meta tx NEEDS to be signed by the wallet owner or one of his guardians in all cases. The reason is that otherwise anyone would be able to completely use up the daily limit of the wallet by creating dummy meta txs on the wallet and paying themselves fees from the wallet. So the relayer can't just create a meta tx by himself for the wallet.

The downside is that the relayer will end up paying the gas costs for the confirmation. I think this is fine as the relayer can request a higher than necessary fee payment for the initial guardian addition. You can think of other solutions but I think this would add unnecessary complexity.

Reply:
Kongliang and I have discussed about this, confirmGuardianAddition and confirmGuardianRemoval now just use useMetaTx=false case, metaTx of others can not pass in current contract as areMetaTxSignersAuthorized check and verifySignatures check(only for wallet owner now).

By keep current design, relay use itself key sign the tx directly call confirmGuardianAddition and confirmGuardianRemoval is a better way, and relay would pay for the gas.

Another way is store the data from owner by using nonce=0 case, but need handle the next add same guardian caused the same hash issue. so the previous way is better.

Reply:
Ah I didn't notice the `useMetaTx = false`. But it does seem like `from: owner` still needs to be changed to a different address (the "relayer").

> but need handle the next add same guardian caused the same hash issue. so the previous way is better.

Yeah, I think we should add a `validUntil` value to `MetaTransaction` to solve this (even if not really needed in this case). This value can be used to limit the validity of the transaction in time, but can also be used to make it unique so the hash is different even if the actual transaction is the same. (I think you could do that as well now by slightly changing some of the other data like `gasLimit` but that would be hacky).